api/src/main/java/org/openmrs/module/registrationcore/api/mpi/pixpdq/PdqSimilarPatientsSearcher.java

```diff
@@ -0,0 +1,145 @@
+import org.openmrs.module.registrationcore.api.search.PatientAndMatchQuality;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+// TODO: Add Javadocs
+public class PdqSimilarPatientsSearcher implements MpiSimilarPatientsSearcher {
+
+    @Autowired
+    @Qualifier("registrationcore.mpiPixPdqMessageUtil")
+    private PixPdqMessageUtil pixPdqMessageUtil;
+
+    @Autowired
+    @Qualifier("registrationcore.Hl7SenderHolder")
+    private Hl7SenderHolder hl7SenderHolder;
+    
+    @Autowired
+    @Qualifier("registrationcore.coreProperties")
+    private RegistrationCoreProperties registrationCoreProperties;
+
+    private final Log log = LogFactory.getLog(this.getClass());
+
+    @Override
+    public List<PatientAndMatchQuality> findSimilarMatches(Patient patient, Map<String, Object> otherDataPoints, Double cutoff, Integer maxResults) {
+        return find(patient, maxResults);
+    }
+
+    @Override
+    public List<PatientAndMatchQuality> findExactMatches(Patient patient, Map<String, Object> otherDataPoints, Double cutoff, Integer maxResults) {
+        return find(patient, maxResults);
+    }
+
+    private List<PatientAndMatchQuality> find(Patient patient, Integer maxResults) {
+
+        List<MpiPatient> mpiPatientList = new LinkedList<MpiPatient>();
+
+        try {
+            List<Map.Entry<String, String>> queryParams = pixPdqMessageUtil.patientToQPD3Params(patient);
+	        if (!queryParams.isEmpty()){
+		        Message pdqRequest = pixPdqMessageUtil.createPdqMessage(queryParams);
+		        Message response = hl7SenderHolder.getHl7v2Sender().sendPdqMessage(pdqRequest);
+		        mpiPatientList = pixPdqMessageUtil.interpretPIDSegments(response);
+	        }
+        } catch (Exception e) {
+            log.error("Error in PDQ Search", e);
+            ErrorHandlingService errorHandler = registrationCoreProperties.getPdqErrorHandlingService();
+            if (errorHandler != null) {
+                errorHandler.handle(e.getMessage(),
+                        "org.openmrs.module.registrationcore.api.mpi.pixpdq.PdqSimilarPatientsSearcher",
+                        true, ExceptionUtils.getFullStackTrace(e));
+            }
+        }
+
+        if (maxResults != null && mpiPatientList.size() > maxResults) {
+            mpiPatientList = mpiPatientList.subList(0, maxResults);
+        }
+        List<? extends Patient> retVal = mpiPatientList;
+        return toMatchList(patient, retVal);
+    }
+
+    // Q: Why doesn't toMatchList check matches for identifiers?
+    private List<PatientAndMatchQuality> toMatchList(Patient patient, List<? extends Patient> patients) {
+        List<PatientAndMatchQuality> matchList = new ArrayList<PatientAndMatchQuality>();
+
+        for (Patient match : patients) {
```

Comment:
Shouldn't the patient matching be happening in the MPI?

Reply:
The matching is done in the MPI to return the patients, however this is done to sort the returned patients to figure out what order the patients should be listed (if I'm not wrong)